### PR TITLE
Fixing handling of multi-source JavaSources on older JDKs.

### PR DESCRIPTION
**Pull request by jlahoda**
The multi-file JavaSources should work reasonably well after https://github.com/apache/netbeans/pull/2959 on new JDKs/with nb-javac, but due to implementation changes in javac, they may not work well on older JDKs (like JDK 14/11). While overall I think we should eventually drop support for older JDKs without (some sort of) nb-javac, that is not the case currently, so trying to make the multi-file sources work also on older JDKs (which should improve the random failures on Travis).

### CODE REVIEW

**Review by sdedic — APPROVED** [review]

**Comment by mbien** [comment]
my understanding was that NB 13 won't have a mode without nb-javac. Its installed by default, based on JDK 17 (backported with jackpot) and will stay enabled even when NB is started on 18+.

**Comment by matthiasblaesing** [comment]
> my understanding was that NB 13 won't have a mode without nb-javac. Its installed by default, based on JDK 17 (backported with jackpot) and will stay enabled even when NB is started on 18+.

Jan fixes a problem that exists now. We can speculate/plan the future, but it has not happened yet: the relevant PRs are #3251 and #2783. They are not merged yet.

**Comment by neilcsmith-net** [comment]
@mbien also note @jlahoda comment in the description -

> While overall I think we should eventually drop support for older JDKs without (some sort of) nb-javac, that is not the case currently,

**Comment by mbien** [comment]
> @mbien also note @jlahoda comment in the description -
> 
> > While overall I think we should eventually drop support for older JDKs without (some sort of) nb-javac, that is not the case currently,

I saw that yes, that was the reason for my comment.

Even if the PRs are not merged yet (it would be a good time to do it soon btw, early in the cycle), I had the impression that the decision was pretty much already made - looks like i was wrong.

File: java/java.source.base/src/org/netbeans/modules/java/source/parsing/JavacParser.java [hunk 1]
```diff
@@ -772,29 +772,25 @@ Phase moveToPhase (final Phase phase, final CompilationInfoImpl currentInfo, Lis
                         compiler.todo.offer(env);
                     }
                 }
-                currentPhase = Phase.RESOLVED;
+                if (!lowMemoryCancel.get()) {
```
[the remark this answers is not in the document]
Comment (jlahoda) [reply java/java.source.base/src/org/netbeans/modules/java/source/parsing/JavacParser.java:775]: Yes, old(er) versions of javac will not propagate the exceptions from the parse and analyze methods. See:
https://github.com/openjdk/jdk/commit/433394203dce55db95caefcb57bac9ec114ebc47

[the remark this answers is not in the document]
Comment (sdedic) [reply java/java.source.base/src/org/netbeans/modules/java/source/parsing/JavacParser.java:775]: @jlahoda Thanks for expanation - I don't have such a wide context, so now I better understand the core piece of this PR :)

**Comment by mbien** [comment]
I noticed both of the above mentioned javac PRs are now integrated. This would make JDK 17 javac the new baseline, right?.

**Review by dbalek — APPROVED** [review]

**Comment by lkishalmi** [comment]
@jlahoda Do we still need this one?

**Comment by neilcsmith-net** [comment]
No reply so rather than bump another milestone, will close. Reopen if still a need.

**Pull request closed without merging**